From 74ff8b387319e6603e85663c196fe31ffff7f2b0 Mon Sep 17 00:00:00 2001
From: Igor Grinberg <igrinberg@habana.ai>
Date: Sun, 20 Jun 2021 20:34:10 +0300
Subject: [PATCH] [SW-45909] fwupd: run fast: add msgs for err cases

Adding error messages for illegal cases makes it easier to identify the
error, understand the flow problems, and prepare a fix.

Change-Id: I478cf39259dab3bce44773e71b4bbb96fe50bb31
Signed-off-by: Igor Grinberg <igrinberg@habana.ai>
---

diff --git a/preboot/module_fwupd.c b/preboot/module_fwupd.c
index 6b4b548..980d747 100644
--- a/preboot/module_fwupd.c
+++ b/preboot/module_fwupd.c
@@ -474,6 +474,7 @@
 static void __module_fwupd_run_fast(struct fwupd_state *state)
 {
 	const char *cmd_str = NULL;
+	const char *err_str = NULL;
 	enum hl_upd_sts sts = HL_UPD_STS_NOOP;
 
 	switch (state->cmd.cmd) {
@@ -496,10 +497,12 @@
 	case HL_UPD_MAP_SPSRAM:
 		cmd_str = cmd_str ?: "map_spsram";
 		/* Protect from multi-MAP */
-		if (state->map_offset || state->map_size)
+		if (state->map_offset || state->map_size) {
 			sts = HL_UPD_STS_VALID_ERR;
-		else
+			err_str = "MAP is already in progress...";
+		} else {
 			sts = HL_UPD_STS_MAP_ACK;
+		}
 
 		module_fwupd_send_status(state, sts, NULL);
 		break;
@@ -509,10 +512,12 @@
 	case HL_UPD_RUN_SPSRAM:
 		cmd_str = cmd_str ?: "run_spsram";
 		/* Protect from RUN before MAP */
-		if (state->map_offset && state->map_size)
+		if (state->map_offset && state->map_size) {
 			sts = HL_UPD_STS_ACK;
-		else
+		} else {
 			sts = HL_UPD_STS_VALID_ERR;
+			err_str = "RUN is not allowed before MAP...";
+		}
 
 		module_fwupd_send_status(state, sts, NULL);
 		break;
@@ -534,6 +539,11 @@
 		preboot_puts_raw(cmd_str);
 		preboot_puts(" received\n");
 	}
+
+	if (err_str) {
+		preboot_puts_raw(err_str);
+		preboot_puts("\n");
+	}
 }
 
 void module_fwupd_run_fast(void)
